scripts/question_answering/similarity_function.py

```diff
@@ -0,0 +1,198 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Collection of general purpose similarity functions"""
+
+import mxnet as mx
+from mxnet import gluon, initializer
+from mxnet.gluon import nn, Parameter
+
+from .utils import combine_tensors
+
+
+class SimilarityFunction(gluon.HybridBlock):
+    """
+    A ``SimilarityFunction`` takes a pair of tensors with the same shape, and computes a similarity
+    function on the vectors in the last dimension.  For example, the tensors might both have shape
+    `(batch_size, sentence_length, embedding_dim)`, and we will compute some function of the two
+    vectors of length `embedding_dim` for each position `(batch_size, sentence_length)`, returning a
+    tensor of shape `(batch_size, sentence_length)`.
+
+    The similarity function could be as simple as a dot product, or it could be a more complex,
+    parameterized function.
+    """
+    default_implementation = 'dot_product'
+
+    def hybrid_forward(self, F, array_1, array_2):  # pylint: disable=arguments-differ
+        # pylint: disable=arguments-differ
+        """
+        Takes two tensors of the same shape, such as ``(batch_size, length_1, length_2,
+        embedding_dim)``.  Computes a (possibly parameterized) similarity on the final dimension
+        and returns a tensor with one less dimension, such as ``(batch_size, length_1, length_2)``.
+        """
+        raise NotImplementedError
+
+
+class DotProductSimilarity(SimilarityFunction):
+    """
+    This similarity function simply computes the dot product between each pair of vectors, with an
+    optional scaling to reduce the variance of the output elements.
+
+    Parameters
+    ----------
+    scale_output : ``bool``, optional
+        If ``True``, we will scale the output by ``F.sqrt(ndarray.shape[-1])``, to reduce the
+        variance in the result.
+    """
+    def __init__(self, scale_output=False, **kwargs):
+        super(DotProductSimilarity, self).__init__(**kwargs)
+        self._scale_output = scale_output
+
+    def hybrid_forward(self, F, array_1, array_2):
+        result = (array_1 * array_2).sum(axis=-1)
+
+        if self._scale_output:
+            result *= F.sqrt(array_1.shape[-1])
+
+        return result
+
+
+class CosineSimilarity(SimilarityFunction):
```

Comment:
This along with several others are available in https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/embedding/evaluation.py#L29

Reply:
Yes, looks like a duplicate, and in this case, other similarity functions should be put in the same place as the rest of them. 

Let's do it once we will move BiDAF and QANet into the master branch.